# AI 헬스케어 데이터 분석 1일차
**주제 : 데이터 다루기의 시작 – CSV 불러오기와 정제**  
**데이터 : `train.csv` (당뇨병 환자 데이터)**

이 노트북은 비전공자도 이해할 수 있도록, 이론 설명 바로 아래에
실습 코드와 작은 연습 문제(미니 문제), 그리고 **정답 예시 코드**를 넣어 구성되어 있습니다.


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


## 학습 목표
1. `pd.read_csv()`로 CSV를 DataFrame으로 불러올 수 있다.  
2. DataFrame의 **행, 열, 인덱스, dtype**을 이해하고 확인할 수 있다.  
3. Pandas의 기본 조작(생성, 선택, 필터링, 정렬, 수정, 추가/삭제)을 실습할 수 있다.  
4. **결측치 & 숨겨진 결측치(0)** 개념을 이해하고, `isnull`, `dropna`, `fillna`로 다룰 수 있다.  
5. Seaborn의 샘플 데이터셋을 불러와 기초 탐색과 간단한 시각화를 할 수 있다.


---
## 데이터·데이터사이언스·데이터엔지니어링 한 줄 정리

- **데이터 분석**: 데이터를 정리·요약·해석해 *무슨 일이 있었고 왜 그런지* 설명하고,  
  의사결정에 바로 쓰일 근거를 만드는 활동입니다.
- **데이터사이언스**: 데이터를 통해 세상의 패턴을 발견하고,  
  미래를 예측하거나 의사결정을 돕는 모델*을 만드는 학문입니다.
- **데이터 엔지니어링 vs 데이터사이언스**
  - 데이터 엔지니어링: 데이터를 **수집·저장·옮기는 파이프라인**을 만듭니다. (도로 깔기)
  - 데이터사이언스: 그 데이터 위에서 **분석·모델링**을 합니다. (도로를 달리기)

오늘은 이 중에서, **데이터사이언스의 기초 도구인 Pandas와 Seaborn**을 연습합니다.


---
## 2-1. Pandas란?

**개념 정리**

- “엑셀을 코딩으로 조작하게 해주는 도구”  
- 표 형태의 데이터를 **데이터프레임(DataFrame)** 구조로 다룸  
- 통계, 필터링, 그룹화, 시각화까지 한 번에 가능  
- 리스트/딕셔너리만으로는 불편한 행·열 단위 분석을,  
  **엑셀 + SQL + 파이썬** 느낌으로 편하게 할 수 있게 해 줌

### 🔹 실습 코드 2-1 – 딕셔너리로 DataFrame 만들기

In [ ]:
import pandas as pd
from IPython.display import display

data = {
    "Name": ["Tom", "Jane", "Alice", "Bob"],
    "Age": [28, 34, 29, 42],
    "Score": [85, 92, 88, 75]
}

df_sample = pd.DataFrame(data)
display(df_sample)
print("\n데이터 타입(dtypes):")
print(df_sample.dtypes)

Name  Age  Score
0    Tom   28     85
1   Jane   34     92
2  Alice   29     88
3    Bob   42     75


데이터 타입(dtypes):
Name     object
Age       int64
Score     int64
dtype: object


### 🔹 미니 문제 2-1
1. 위 `df_sample`에서 행(Row)은 몇 개이고, 열(Column)은 몇 개인가요?  
2. `Score` 컬럼의 데이터 타입(dtype)은 무엇인가요?


In [ ]:
print(df_sample.shape)

(4, 3)


In [ ]:
# 💡 미니 문제 2-1 정답 예시
rows, cols = df_sample.shape
print("행 개수:", rows)
print("열 개수:", cols)
print("Score 컬럼 dtype:", df_sample["Score"].dtype)

행 개수: 4
열 개수: 3
Score 컬럼 dtype: int64


---
## 2-2. 데이터프레임 기본 구조 & 인덱스 조작

**구성요소**

| 구성요소 | 설명                     |
|----------|--------------------------|
| 행(Row)  | 개별 데이터 (예: 한 환자) |
| 열(Column) | Feature / 변수           |
| dtype    | 데이터 타입 (int, float, object 등) |

**인덱스 조작**
- 기본 인덱스는 0, 1, 2, …  
- `df.index = ["a", "b", "c", "d"]` 처럼 직접 이름을 줄 수도 있음  
- 인덱스를 바꾸면 행을 구분하기 쉬워짐

### 🔹 실습 코드 2-2 – 인덱스 바꾸고, 컬럼/행 선택해 보기

In [ ]:
data = {
    "Name": ["Tom", "Jane", "Alice", "Bob"],
    "Age": [28, 34, 29, 42],
    "Score": [85, 92, 88, 75]
}
df_sample = pd.DataFrame(data)

In [ ]:
# 인덱스 변경
df_sample.index = ["a", "b", "c", "d"]
print("=== 인덱스 변경 후 ===")
display(df_sample)

=== 인덱스 변경 후 ===


Name  Age  Score
a    Tom   28     85
b   Jane   34     92
c  Alice   29     88
d    Bob   42     75

In [ ]:
# 단일 컬럼 선택 (시리즈)
print("=== Name 컬럼 (Series) ===")
display(df_sample["Name"])

=== Name 컬럼 (Series) ===


a      Tom
b     Jane
c    Alice
d      Bob
Name: Name, dtype: object

In [ ]:
# 여러 컬럼 선택 (데이터프레임)
print("=== Name, Score 컬럼 (DataFrame) ===")
display(df_sample[["Name", "Score"]])

=== Name, Score 컬럼 (DataFrame) ===


Name  Score
a    Tom     85
b   Jane     92
c  Alice     88
d    Bob     75

In [ ]:
# 행 선택: 인덱스 이름 (loc)
print("=== 인덱스 'a' 행 (loc) ===")
display(df_sample.loc["a"])

=== 인덱스 'a' 행 (loc) ===


Name     Tom
Age       28
Score     85
Name: a, dtype: object

In [ ]:
# 행 선택: 행 번호 (iloc)
print("=== 0번째 행 (iloc) ===")
display(df_sample.iloc[0])

=== 0번째 행 (iloc) ===


Name     Tom
Age       28
Score     85
Name: a, dtype: object

### 🔹 미니 문제 2-2
1. 인덱스를 `["학생1", "학생2", "학생3", "학생4"]`로 바꾸려면 어떤 코드를 쓰면 될까요?  
2. `Age`와 `Score`만 포함된 새로운 DataFrame을 한 줄 코드로 만들어 보세요.


In [ ]:
# 💡 미니 문제 2-2 정답 예시
# 1) 인덱스 변경
df_sample.index = ["학생1", "학생2", "학생3", "학생4"]
print(df_sample)

# 2) Age와 Score만 포함된 새로운 DataFrame
df_age_score = df_sample[["Age", "Score"]]
print("\nAge & Score만 포함된 DataFrame:")
print(df_age_score)

      Name  Age  Score
학생1    Tom   28     85
학생2   Jane   34     92
학생3  Alice   29     88
학생4    Bob   42     75

Age & Score만 포함된 DataFrame:
     Age  Score
학생1   28     85
학생2   34     92
학생3   29     88
학생4   42     75


---
## 2-3. Pandas 기본 사용법 – 조건, 정렬, 수정, 추가/삭제

**자주 쓰는 기본 조작들**

- 행 조건 필터링: `df[df["Score"] > 80]`  
- 여러 조건 결합: `df[(df["Score"] > 80) & (df["Age"] < 35)]`  
- 정렬: `df.sort_values(by="Score", ascending=False)`  
- 값 수정: `df.loc[df["Name"] == "Tom", "Score"] = 90`  
- 새로운 컬럼 추가: `df["Passed"] = df["Score"] >= 80`  
- 컬럼 삭제: `df.drop(columns=["Passed"], inplace=True)`  
- 행 삭제: `df.drop(index="b", inplace=True)`

### 🔹 실습 코드 2-3 – Pandas 기본 조작 종합 연습

In [ ]:
data = {
    "Name": ["Tom", "Jane", "Alice", "Bob"],
    "Age": [28, 34, 29, 42],
    "Score": [85, 92, 88, 75]
}
df_sample = pd.DataFrame(data)
df_sample.index = ["a", "b", "c", "d"]
print("=== 원본 데이터 ===")
display(df_sample)

=== 원본 데이터 ===


Name  Age  Score
a    Tom   28     85
b   Jane   34     92
c  Alice   29     88
d    Bob   42     75

=== Score > 80인 학생 ===


Name  Age  Score
a    Tom   28     85
b   Jane   34     92
c  Alice   29     88

=== Score > 80 AND Age < 35 ===


Name  Age  Score
a    Tom   28     85
b   Jane   34     92
c  Alice   29     88

=== Score 기준 내림차순 정렬 ===


Name  Age  Score
b   Jane   34     92
c  Alice   29     88
a    Tom   28     85
d    Bob   42     75

=== Tom의 점수 수정 후 ===


Name  Age  Score
a    Tom   28     90
b   Jane   34     92
c  Alice   29     88
d    Bob   42     75

=== Passed 컬럼 추가 후 ===


Name  Age  Score  Passed
a    Tom   28     90    True
b   Jane   34     92    True
c  Alice   29     88    True
d    Bob   42     75   False

=== Passed 컬럼 삭제 후 ===


Name  Age  Score
a    Tom   28     90
b   Jane   34     92
c  Alice   29     88
d    Bob   42     75

=== 인덱스 'b' 행 삭제 후 ===


Name  Age  Score
a    Tom   28     90
c  Alice   29     88
d    Bob   42     75

In [ ]:

# 1) 점수가 80점 초과인 학생
print("=== Score > 80인 학생 ===")
display(df_sample[df_sample["Score"] > 80])

In [ ]:
# 2) Score > 80 AND Age < 35
print("=== Score > 80 AND Age < 35 ===")
display(df_sample[(df_sample["Score"] > 80) & (df_sample["Age"] < 35)])

In [ ]:
# 3) Score 기준 내림차순 정렬
print("=== Score 기준 내림차순 정렬 ===")
display(df_sample.sort_values(by="Score", ascending=False))

In [ ]:
# 4) Tom의 Score를 90점으로 수정
df_sample.loc[df_sample["Name"] == "Tom", "Score"] = 90
print("=== Tom의 점수 수정 후 ===")
display(df_sample)

In [ ]:
# 5) Passed 컬럼 추가 (Score >= 80)
df_sample["Passed"] = df_sample["Score"] >= 80
print("=== Passed 컬럼 추가 후 ===")
display(df_sample)

In [ ]:
# 6) Passed 컬럼 삭제
df_sample.drop(columns=["Passed"], inplace=True)
print("=== Passed 컬럼 삭제 후 ===")
display(df_sample)

In [ ]:
# 7) 인덱스 'b' 행 삭제
df_sample.drop(index="b", inplace=True)
print("=== 인덱스 'b' 행 삭제 후 ===")
display(df_sample)

### 🔹 미니 문제 2-3
1. `Score`가 **90점 이상**인 학생만 필터링하는 코드를 직접 작성해 보세요.  
2. 새 컬럼 `Level`을 만들어, `Score`가 85점 이상이면 `'High'`, 그렇지 않으면 `'Normal'`을 넣으려면  
   `df.loc[조건, "Level"] = 값` 형태를 활용해 어떤 코드를 작성할 수 있을까요?


In [ ]:
# 💡 미니 문제 2-3 정답 예시
data = {
    "Name": ["Tom", "Jane", "Alice", "Bob"],
    "Age": [28, 34, 29, 42],
    "Score": [90, 92, 88, 75]
}
df_q = pd.DataFrame(data)
print("원본:")
print(df_q)

In [ ]:
# 1) Score가 90점 이상인 학생
high_scorers = df_q[df_q["Score"] >= 90]
print("\nScore >= 90 인 학생:")
print(high_scorers)

In [ ]:
# 2) Level 컬럼 추가
df_q.loc[df_q["Score"] >= 85, "Level"] = "High"
df_q.loc[df_q["Score"] < 85, "Level"] = "Normal"
print("\nLevel 컬럼 추가 결과:")
print(df_q)

---
## 2-4. CSV로 된 당뇨병 데이터 불러오기

이제 실제 **당뇨병 환자 데이터(train.csv)**를 불러와 보겠습니다.

### 🔹 실습 코드 3 – CSV 파일 불러오기 & 구조 확인

In [ ]:
import pandas as pd
from IPython.display import display

df = pd.read_csv("/content/train (1).csv")

print("데이터 크기 (행, 열):", df.shape)
print("\n=== head() : 위에서 5행 ===")
display(df.head())

print("\n=== info() : 컬럼별 타입 / 결측 여부 ===")
df.info()

print("\n=== dtypes : 데이터 타입만 ===")
print(df.dtypes)

데이터 크기 (행, 열): (652, 10)

=== head() : 위에서 5행 ===


ID  Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin  \
0  TRAIN_000            4      103             60             33      192   
1  TRAIN_001           10      133             68              0        0   
2  TRAIN_002            4      112             78             40        0   
3  TRAIN_003            1      119             88             41      170   
4  TRAIN_004            1      114             66             36      200   

    BMI  DiabetesPedigreeFunction  Age  Outcome  
0  24.0                     0.966   33        0  
1  27.0                     0.245   36        0  
2  39.4                     0.236   38        0  
3  45.3                     0.507   26        0  
4  38.1                     0.289   21        0


=== info() : 컬럼별 타입 / 결측 여부 ===
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 652 entries, 0 to 651
Data columns (total 10 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   ID                        652 non-null    object 
 1   Pregnancies               652 non-null    int64  
 2   Glucose                   652 non-null    int64  
 3   BloodPressure             652 non-null    int64  
 4   SkinThickness             652 non-null    int64  
 5   Insulin                   652 non-null    int64  
 6   BMI                       652 non-null    float64
 7   DiabetesPedigreeFunction  652 non-null    float64
 8   Age                       652 non-null    int64  
 9   Outcome                   652 non-null    int64  
dtypes: float64(2), int64(7), object(1)
memory usage: 51.1+ KB

=== dtypes : 데이터 타입만 ===
ID                           object
Pregnancies                   int64
Glucose                       int64
Blood

### 🔹 미니 문제 3
1. `ID` 컬럼의 데이터 타입(dtype)은 무엇인가요?  
2. 숫자 연산(평균, 합, 그래프 등)에 주로 사용할 컬럼 3개를 골라 보세요. (예: Glucose, BMI, Age …)


In [ ]:
# 💡 미니 문제 3 정답


np.int64(0)

In [ ]:
df.isnull().sum(axis=1)

0      0
1      0
2      0
3      0
4      0
      ..
647    0
648    0
649    0
650    0
651    0
Length: 652, dtype: int64

---
## 2-5. 기초 탐색 – `head`, `shape`, `describe`

### 🔹 실습 코드 4 – 분포와 이상한 값(0) 감지하기

In [ ]:
display(df.describe())

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   652.000000  652.000000     652.000000     652.000000  652.000000   
mean      3.923313  120.946319      68.895706      20.440184   80.878834   
std       3.327529   31.992394      19.183843      15.918604  117.748315   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      64.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   34.000000   
75%       6.000000  142.000000      80.000000      32.000000  126.000000   
max      15.000000  199.000000     114.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  652.000000                652.000000  652.000000  652.000000  
mean    31.947393                  0.465133   33.510736    0.349693  
std      7.460975                  0.329496   11.824817    0.477239  
min      0.000000                  0.084000   21.000000    0.000000  
25%     27.400000                  0.238750   24.000000    0.000000  
50%     32.000000                  0.363000   29.500000    0.000000  
75%     36.425000                  0.607750   41.000000    1.000000  
max     57.300000                  2.329000   81.000000    1.000000

### 🔹 미니 문제 4
1. `Glucose`(혈당) 컬럼의 `min` 값은 몇인가요?  
2. 그 값이 실제 사람에게서 나올 수 있는 값인지, 여러분의 생각을 적어보세요.


In [ ]:
# 💡 미니 문제 4 정답
glucosr_min = df["Glucose"].min()
glucosr_min
#사람의 혈당이 혈당이 0이 나오는 것은 아마 측정값이 누락된 것 같다.

0

---
## 2-6. 결측치(Missing Value)와 숨겨진 결측치

### 🔹 실습 코드 5 – 표면상 결측치 확인 (`isnull().sum()`)

In [ ]:
print(df.isnull().sum())

ID                            0
Pregnancies                   0
Glucose                       4
BloodPressure                30
SkinThickness               195
Insulin                     318
BMI                           7
DiabetesPedigreeFunction      0
Age                           0
Outcome                       0
dtype: int64


---
## 겉으로 보이는 결측치(빈 칸) 먼저 확인하기

### 코드
```python
print(df.isnull().sum())
```

### 설명 (비유)
- `df`는 "당뇨병 환자 정보가 들어 있는 엑셀 표"라고 생각하세요.
- `isnull()`은 **각 칸이 비어 있는지** 물어보는 함수입니다.
  - 비어 있으면 `True`
  - 값이 있으면 `False`
- `.sum()`은 `True`를 1로 보고, **열(컬럼)별로 더합니다.**
  - 결과: 각 컬럼마다 `비어 있는 칸 개수`가 나옵니다.

즉, 이 코드는 이렇게 읽을 수 있습니다.

> "각 항목(혈당, 혈압, BMI 등)에 **빈 칸이 몇 개나 있는지** 세어줘."

### 🔹 미니 문제 5
1. 결과를 보면 모든 컬럼의 결측치 개수가 0입니다. 정말로 완벽한 데이터라고 볼 수 있을까요?  
2. "숨겨진 결측치" 관점에서 어떤 걸 추가로 의심해 봐야 할까요?


In [ ]:
# 💡 미니 문제 5 정답
#1. 아니요
#2. 0을 평균값으로 채워 넣는다.

### 🔹 실습 코드 6 – 0값(숨겨진 결측 후보) 개수 확인

In [ ]:
potential_zeros = ["Glucose", "BloodPressure", "SkinThickness", "Insulin", "BMI"]
zero_counts = (df[potential_zeros] == 0).sum()
print(zero_counts)

Glucose            4
BloodPressure     30
SkinThickness    195
Insulin          318
BMI                7
dtype: int64


---
## 의학적으로 0이 이상한 값(숨겨진 결측치 후보) 개수 세기

### 코드
```python
potential_zeros = ["Glucose", "BloodPressure", "SkinThickness", "Insulin", "BMI"]
zero_counts = (df[potential_zeros] == 0).sum()
print(zero_counts)
```

### 설명
의료 데이터에서는 값이 비었을 때 **진짜 빈 칸** 대신 `0`으로 적어놓는 경우가 많습니다.

예를 들어,
- 혈당(Glucose) = 0 → 실제로는 거의 불가능한 값
- 혈압(BloodPressure) = 0 → 사람이 살아 있을 수 없음

그래서 이런 0은 **'측정을 못 해서 비어 있는데, 0으로 기록해 둔 값'**일 가능성이 큽니다.

- `potential_zeros = [...]` : 0이 나오면 의심되는 컬럼 목록을 적어둡니다.
- `df[potential_zeros] == 0` : 이 컬럼들에서 값이 0인 칸은 `True`, 나머지는 `False`가 됩니다.
- `.sum()` : `True` 개수를 세어, 각 컬럼별로 `0의 개수`를 구합니다.

즉, 이 코드는 이렇게 해석할 수 있습니다.

> "혈당, 혈압, 피부두께, 인슐린, BMI에서 **0으로 기록된 칸이 몇 개인지** 세어줘."

### 🔹 미니 문제 6
1. `Insulin` 컬럼에서 0이 몇 개 있나요?  
2. 이 0들을 그냥 “정상 데이터”로 두면, 나중에 어떤 문제가 생길 수 있을까요?


In [ ]:
potential_zeros = ["Glucose", "BloodPressure", "SkinThickness", "Insulin", "BMI"]
zero_counts = (df[potential_zeros] == 0).sum()
print(zero_counts)

Glucose            4
BloodPressure     30
SkinThickness    195
Insulin          318
BMI                7
dtype: int64


In [ ]:
# 💡 미니 문제 6 정답
insiln_zero = (df["Insulin"] == 0).sum()
insiln_zero
#2. 정상이 나오지 않고 이상한 값이 나온다. 그래서 나중에 잘못된 처방이 될 수 있다.

np.int64(318)

---
## 인슐린이 0인 환자 수만 따로 뽑아보기

### 코드
```python
insulin_zero = zero_counts["Insulin"]
```

### 설명
위에서 `zero_counts`는 대략 이런 모양입니다.

```text
Glucose           4
BloodPressure    30
SkinThickness   195
Insulin         318
BMI               7
```

`zero_counts["Insulin"]`은 그 중에서 **Insulin 행만 딱 골라서** 숫자를 가져오는 것입니다.
즉, `insulin_zero`에는 “인슐린 값이 0으로 기록된 환자 수”가 들어갑니다.

보고서에서는 이렇게 쓸 수 있겠죠.

> "전체 652명 중, 인슐린 값이 0으로 기록된 사람은 318명입니다."

---
## 2-7. 0을 NaN으로 바꾸고, 결측치 처리하기

### 1) 0 → NaN으로 변환 (`replace`)

### 🔹 실습 코드 7-1 -> 0을 NaN으로 변환하고 결측 개수 다시 보기

In [ ]:
import numpy as np
# 💡 미니 문제 7-1 정답
# zero_counts.replace(0, np.nan, inplace=True)
# print(zero_counts)
df[potential_zeros] = df[potential_zeros].replace(0, np.nan)


---
## 의심스러운 0을 진짜 결측치(NaN)로 바꾸기

### 코드
```python
import numpy as np
potential_zeros = ["Glucose", "BloodPressure", "SkinThickness", "Insulin", "BMI"]
df[potential_zeros] = df[potential_zeros].replace(0, np.nan)
```

### 설명
이제 0을 **"진짜 없는 값"**으로 처리하려고 합니다.

- `np.nan` 은 "값 없음"을 나타내는 특수한 표시입니다.
- `df[potential_zeros].replace(0, np.nan)` 은
  - 지정한 컬럼에서 `0`을 모두 `NaN`으로 바꿔줍니다.
- 마지막에 `df[potential_zeros] = ...` 로 다시 덮어쓰면서,
  - 이 변경 사항을 `df`에 반영합니다.

의미를 문장으로 쓰면:

> "혈당/혈압/피부두께/인슐린/BMI에서 **0으로 적혀 있던 값들을, 이제부터는 '빈 값(NaN)'으로 인정하겠다."

### 🔹 미니 문제 7-2
1. `Glucose` 결측치 개수는 몇 개가 되었나요?  
2. 왜 0을 NaN으로 바꾸는 작업이 중요한지, 한 줄로 정리해 보세요.


In [ ]:
# 💡 미니 문제 7-2 정답 예시
df["Glucose"].isnull().sum()
# 2.바꿔야하는 이유는 0의 값이 있으면 이상한값이 나오고 NaN는 다른 것으로 대체를 하거나 제거시켜면 되기떄문이다.

np.int64(4)

---
## Glucose(혈당)에서 결측치 개수 다시 세기

### 코드
```python
glucose_missing = df["Glucose"].isnull().sum()
```

### 설명
0을 NaN으로 바꾸고 나면, 이제 Glucose 컬럼에는 진짜 결측치(NaN)가 생깁니다.

- `df["Glucose"].isnull()` : 혈당 칸이 비어 있으면 `True`, 아니면 `False`
- `.sum()` : `True` 개수 = 결측치 개수

즉, 이 코드는:

> "혈당(Glucose) 값이 비어 있는(또는 0에서 바뀐) 환자가 **몇 명인지** 세어줘."

### 2) 결측치 제거 (`dropna`)

### 🔹 실습 코드 8-1 – Glucose, BloodPressure 결측 행 삭제

In [ ]:
# 💡 미니 문제 8-1 정답
df.dropna(subset =['Glucose','BloodPressure'])

ID  Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin  \
0    TRAIN_000            4    103.0           60.0           33.0    192.0   
1    TRAIN_001           10    133.0           68.0            NaN      NaN   
2    TRAIN_002            4    112.0           78.0           40.0      NaN   
3    TRAIN_003            1    119.0           88.0           41.0    170.0   
4    TRAIN_004            1    114.0           66.0           36.0    200.0   
..         ...          ...      ...            ...            ...      ...   
647  TRAIN_647            1     91.0           64.0           24.0      NaN   
648  TRAIN_648           10    122.0           68.0            NaN      NaN   
649  TRAIN_649            8     84.0           74.0           31.0      NaN   
650  TRAIN_650            2     81.0           72.0           15.0     76.0   
651  TRAIN_651            1    107.0           68.0           19.0      NaN   

      BMI  DiabetesPedigreeFunction  Age  Outcome  
0    24.0                     0.966   33        0  
1    27.0                     0.245   36        0  
2    39.4                     0.236   38        0  
3    45.3                     0.507   26        0  
4    38.1                     0.289   21        0  
..    ...                       ...  ...      ...  
647  29.2                     0.192   21        0  
648  31.2                     0.258   41        0  
649  38.3                     0.457   39        0  
650  30.1                     0.547   25        0  
651  26.5                     0.165   24        0  

[618 rows x 10 columns]

아래 내용 그대로 코랩 마크다운(텍스트) 셀에 붙여 넣어서 쓰면 됩니다 😊

---

### 코드 설명

```python
before_rows = len(df)
df_drop = df.dropna(subset=["Glucose", "BloodPressure"])
after_rows = len(df_drop)
print("삭제 전 행 개수:", before_rows)
print("삭제 후 행 개수:", after_rows)
print("삭제된 행 개수:", before_rows - after_rows)
```

위 코드는 **결측치(NA)가 있는 행을 제거한 뒤, 그 전·후 행 개수를 비교**하는 과정이다.

1. `before_rows = len(df)`

   * 원본 데이터프레임 `df`가 가지고 있는 **전체 행(row) 개수**를 구해서 `before_rows` 변수에 저장한다.
   * 즉, **삭제 전 행 개수**를 기록하는 단계다.

2. `df_drop = df.dropna(subset=["Glucose", "BloodPressure"])`

   * `Glucose`와 `BloodPressure` 두 컬럼을 기준으로,
     이 컬럼들 중 **하나라도 결측치(NA)가 있는 행을 삭제**한다.
   * 그 결과가 `df_drop`이라는 새로운 데이터프레임에 저장된다.
   * 원본 `df`는 그대로 두고, **결측치를 제거한 버전이 `df_drop`** 이다.

3. `after_rows = len(df_drop)`

   * 결측치 제거 후 데이터프레임 `df_drop`의 **행 개수**를 구해서 `after_rows` 변수에 저장한다.
   * 즉, **삭제 후 남은 행 개수**를 기록하는 단계다.

4. ```python
   print("삭제 전 행 개수:", before_rows)
   print("삭제 후 행 개수:", after_rows)
   print("삭제된 행 개수:", before_rows - after_rows)
   ```

   * 첫 번째 줄: 원본 데이터 기준 **삭제 전 행 개수** 출력
   * 두 번째 줄: 결측치를 제거한 **삭제 후 행 개수** 출력
   * 세 번째 줄: 두 값을 빼서

     * **실제로 몇 개의 행이 삭제되었는지(결측치 때문에 제거된 행 수)** 를 출력

이 코드를 통해,

> “Glucose와 BloodPressure에 결측치가 있는 데이터가 얼마나 되는지”
> 를 **삭제된 행 개수**로 직관적으로 확인할 수 있다.


### 🔹 미니 문제 8-2
1. Glucose와 BloodPressure 결측으로 인해 삭제된 행(환자)은 몇 명인가요?  
2. 만약 전체의 절반 이상이 삭제된다면, 이 방법을 그대로 쓰는 것이 좋을까요?


In [ ]:
# 💡 미니 문제 8-2 정답 예시
# 1. 34명
# 2. fillna를ㄹ 써서 평균으로 채운다

### 3) 결측치 채우기 (`fillna`)와 평균/중앙값/최빈값

### 🔹 실습 코드 9 – 숫자형 컬럼 결측을 평균으로 채우기

In [ ]:
numeric_cols = df.select_dtypes(include=["int64", "float64"]).columns
df[numeric_cols] = df[numeric_cols].fillna(df[numeric_cols].mean())
print(df.isnull().sum())

ID                          0
Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64


---
## 숫자형 컬럼 결측치를 평균으로 채우기

### 코드 (1단계)
```python
numeric_cols = df.select_dtypes(include=["int64", "float64"]).columns
df[numeric_cols] = df[numeric_cols].fillna(df[numeric_cols].mean())
print(df.isnull().sum())
```

### 설명
- `select_dtypes(include=["int64", "float64"])` : 숫자형 컬럼(정수, 실수)만 골라냅니다.
- `df[numeric_cols].fillna(df[numeric_cols].mean())`
  - 숫자형 컬럼들에서 비어 있는 칸(NaN)을, 그 컬럼의 **평균값**으로 채웁니다.
  - 예: BMI가 비어 있는 사람 → 전체 BMI 평균 값으로 채우기
- 마지막에 `print(df.isnull().sum())` 으로 결측치가 잘 채워졌는지 다시 확인합니다.

이 단계는 이렇게 이해할 수 있습니다.

> "숫자형 항목에서 비어 있는 값들은, **그 항목의 평균값으로 적당히 메꿔서 분석을 진행하자.**"

### 🔹 미니 문제 9
1. `Glucose` 평균으로 결측을 채우는 것은 어떤 장점과 단점을 갖고 있을까요?  
2. “의료 데이터”라는 관점에서, 평균 대신 어떤 값(중앙값, 정상 범위의 대표값 등)을 쓸 수도 있을지 생각해 보세요.


In [ ]:
# 💡 미니 문제 9 정답 예시 (설명 출력)
# 1. 평균값으로 채우면 코드가 간략화, 단점은 이상치 떄문에 값이 이상해짐
# 2. 정상값이나 중앙값을 사용하면 좋을 것 같다.

---
## 3. 실무형 미션 – 당뇨병 데이터로 업무 요청 해결하기

아래 미션들은 전처리 완료된 `df`를 사용합니다.

### [공통 준비 셀] – 전처리 완료 DataFrame 만들기

In [ ]:
import numpy as np
import pandas as pd
from IPython.display import display
import matplotlib.pyplot as plt

#1. csv파일 불러오기
df = pd.read_csv("/content/train (1).csv") # DataFrame

#2. 의료 데이터에서 0이 될 수 없어서 '숨겨진 결측치'로 불 수 있는 컬럼 목록

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 652 entries, 0 to 651
Data columns (total 10 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   ID                        652 non-null    object 
 1   Pregnancies               652 non-null    int64  
 2   Glucose                   652 non-null    int64  
 3   BloodPressure             652 non-null    int64  
 4   SkinThickness             652 non-null    int64  
 5   Insulin                   652 non-null    int64  
 6   BMI                       652 non-null    float64
 7   DiabetesPedigreeFunction  652 non-null    float64
 8   Age                       652 non-null    int64  
 9   Outcome                   652 non-null    int64  
dtypes: float64(2), int64(7), object(1)
memory usage: 51.1+ KB


In [ ]:
potential_zeros = ["Glucose", "BloodPressure", "SkinThickness", "Insulin", "BMI"]
df[potential_zeros] = df[potential_zeros].replace(0,np.nan)

#4. Glucose, BloodPressure 값이 NaN인 행(환자)를 분석해서 제외
# 사유 : 이 두 자료가 없는 환자는 품질이 낮다고 판단하여 드랍
df = df.dropna(subset=["Glucose", "BloodPressure"])

#5. 숫자형 컬럼들만 선택(int64, floast64 타입)
numeric_cols = df.select_dtypes(include=["int64", "float64"]).columns

#6. 숫자형 컬럼들이 남아 있는 NaN을 각 평균값으로 채우기
df = df.fillna(df.mean()["SkinThickness", "Insulin", "BMI"])

#7. 최종적으로 결측치가 모두 정리되었있는 거확인
df.isnull().sum()

#8. 전처리된 데이터 상위 5행 회면에 출력
df.head()

<class 'pandas.core.frame.DataFrame'>
Index: 618 entries, 0 to 651
Data columns (total 10 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   ID                        618 non-null    object 
 1   Pregnancies               618 non-null    int64  
 2   Glucose                   618 non-null    float64
 3   BloodPressure             618 non-null    float64
 4   SkinThickness             451 non-null    float64
 5   Insulin                   333 non-null    float64
 6   BMI                       616 non-null    float64
 7   DiabetesPedigreeFunction  618 non-null    float64
 8   Age                       618 non-null    int64  
 9   Outcome                   618 non-null    int64  
dtypes: float64(6), int64(3), object(1)
memory usage: 53.1+ KB


In [ ]:
df = df.fillna(df.mean()["SkinThickness", "Insulin", "BMI"])

### 미션 1. “평균 혈당 보고서 부탁드려요”

**업무 요청**

> “전체 환자의 평균 혈당이 어느 정도인지 알고 싶습니다.  
> 그리고 당뇨병 환자(Outcome=1)와 비당뇨병 환자(Outcome=0)의 평균 혈당 차이도 같이 알려주세요.”


In [ ]:
glucose_mean = df["Glucose"].mean()
print(f"전체 환자 평균 혈당 : {glucose_mean}")
mean_glucose = df.groupby('Outcome')['Glucose'].mean()
print("Outcome별 평균 혈당 : ")
print(mean_glucose)


전체 환자 평균 혈당 : 121.79449838187702
Outcome별 평균 혈당 : 
Outcome
0    111.113022
1    142.398104
Name: Glucose, dtype: float64



---

### 코드 설명

```python
overall_glucose_mean = df["Glucose"].mean()
grouped_glucose = df.groupby("Outcome")["Glucose"].mean()
print(f"전체 평균 혈당(Glucose): {overall_glucose_mean:.2f}\n")
print("Outcome별 평균 혈당:")
print(grouped_glucose)
```

이 코드는 **전체 데이터의 혈당 평균**과 **Outcome(예: 0/1)에 따른 혈당 평균**을 각각 구해서 비교하는 코드이다.
(Outcome이 0 = 질병 없음, 1 = 질병 있음 같은 라벨이라고 가정)

---

#### 1. 전체 평균 혈당 구하기

```python
overall_glucose_mean = df["Glucose"].mean()
```

* 데이터프레임 `df`에서 `Glucose` 컬럼만 선택한 뒤,
* 그 값들의 **평균(mean)** 을 계산한다.
* 계산된 결과(전체 데이터의 혈당 평균)를 `overall_glucose_mean` 변수에 저장한다.
* 즉, **“전체 사람들의 평균 혈당이 얼마인지”**를 구하는 단계이다.

---

#### 2. Outcome별(그룹별) 평균 혈당 구하기

```python
grouped_glucose = df.groupby("Outcome")["Glucose"].mean()
```

* `df.groupby("Outcome")`
  → `Outcome` 값(예: 0, 1) 기준으로 데이터를 **그룹 나누기**
* `["Glucose"]`
  → 각 그룹에서 `Glucose` 컬럼만 선택
* `.mean()`
  → 각 그룹에 대해 `Glucose`의 **평균값**을 계산
* 결과적으로 `grouped_glucose`에는:

  * Outcome = 0인 그룹의 평균 혈당
  * Outcome = 1인 그룹의 평균 혈당
    이 **시리즈(Series)** 형태로 저장된다.

즉,
**“질병 없음(0) 그룹 vs 질병 있음(1) 그룹의 평균 혈당을 비교”**할 수 있게 만들어 주는 코드이다.

---

#### 3. 전체 평균 혈당 출력

```python
print(f"전체 평균 혈당(Glucose): {overall_glucose_mean:.2f}\n")
```

* `f` 문자열 포매팅을 사용해서:

  * `overall_glucose_mean` 값을 **소수 둘째 자리까지(`:.2f`)** 출력한다.
* `"전체 평균 혈당(Glucose): ..."` 라는 문구와 함께 보여 줌으로써
  **전체 데이터의 혈당 수준이 어느 정도인지** 직관적으로 확인할 수 있다.
* `\n` 은 줄바꿈 문자로, 다음 출력과 사이에 한 줄 띄우기 위한 용도이다.

---

#### 4. Outcome별 평균 혈당 출력

```python
print("Outcome별 평균 혈당:")
print(grouped_glucose)
```

* 첫 줄에서 안내 문구 `"Outcome별 평균 혈당:"` 을 출력하고,
* 다음 줄에서 `grouped_glucose` 값을 출력한다.

  * 예:

    * Outcome 0 → 평균 혈당 120.xx
    * Outcome 1 → 평균 혈당 140.xx
      같은 식으로, 라벨별 평균값이 나온다.
* 이를 통해,

  * **Outcome에 따라 혈당 수준이 차이가 나는지**,
  * 즉, **혈당이 질병 여부와 어느 정도 관련이 있어 보이는지**를 탐색할 수 있다.

---

### 전체 흐름 한 줄 요약

> 이 코드는
> ① 전체 데이터의 평균 혈당을 구하고,
> ② Outcome 값(0/1)에 따라 그룹을 나눠 각 그룹의 평균 혈당을 구한 뒤,
> ③ 두 결과를 출력해서 **“전체 vs 그룹별” 혈당 수준을 비교**하기 위한 코드이다.


### 미션 2. “당뇨 환자 연령대 요약 좀 해주세요”

**업무 요청**

> “당뇨 환자(Outcome=1)의 최소/최대/평균/중앙값 나이를 알고 싶습니다.”


In [ ]:
outcome_pa = df[df["Outcome"] == 1]
outcome_pa_min = outcome_pa["Age"].min()
outcome_pa_max = outcome_pa["Age"].max()
outcome_pa_mean = outcome_pa["Age"].mean()
outcome_pa_median = outcome_pa["Age"].median()


print(f"최솟값 : {outcome_pa_min}")
print(f"최댓값 : {outcome_pa_max}")
print(f"평균값 : {outcome_pa_mean}")
print(f"중앙값 : {outcome_pa_median}")

최솟값 : 21
최댓값 : 70
평균값 : 37.36492890995261
중앙값 : 36.0



---

### 코드 설명

```python
diabetic = df[df["Outcome"] == 1]
age_min = diabetic["Age"].min()
age_max = diabetic["Age"].max()
age_mean = diabetic["Age"].mean()
age_median = diabetic["Age"].median()
print("=== 당뇨 환자(Outcome=1) 연령 요약 ===")
print(f"최소 나이 : {age_min}")
print(f"최대 나이 : {age_max}")
print(f"평균 나이 : {age_mean:.2f}")
print(f"중앙값   : {age_median}")
```

이 코드는 **당뇨 환자(Outcome = 1)만 따로 뽑아서, 그들의 나이에 대한 기초 통계값(최소, 최대, 평균, 중앙값)을 요약해서 출력하는 코드**이다.

---

#### 1. 당뇨 환자(Outcome=1)만 필터링

```python
diabetic = df[df["Outcome"] == 1]
```

* `df["Outcome"] == 1`
  → Outcome 값이 1인 행만 `True`가 되고 나머지는 `False`가 된다.
* `df[조건]` 형태로 쓰면, 조건이 `True`인 행만 남긴 새로운 데이터프레임이 만들어진다.
* 따라서 `diabetic`에는
  → **Outcome이 1인 사람들(당뇨 환자로 분류된 사람들)만 모인 데이터프레임** 이 저장된다.

---

#### 2. 최소 나이 구하기

```python
age_min = diabetic["Age"].min()
```

* `diabetic` 데이터프레임에서 `Age` 컬럼만 선택한 뒤,
* 그 값들 중 **가장 작은 값(최소값)** 을 계산한다.
* 즉, 당뇨 환자들 중에서 **가장 어린 사람의 나이**를 의미한다.

---

#### 3. 최대 나이 구하기

```python
age_max = diabetic["Age"].max()
```

* `Age` 컬럼에서 **가장 큰 값(최대값)** 을 구한다.
* 즉, 당뇨 환자들 중에서 **가장 나이가 많은 사람의 나이**를 의미한다.

---

#### 4. 평균 나이 구하기

```python
age_mean = diabetic["Age"].mean()
```

* `Age` 값들의 **평균(산술 평균)** 을 계산한다.
* 당뇨 환자들의 “**전체적으로 평균적인 나이**”가 어느 정도인지 파악할 수 있다.

---

#### 5. 중앙값 구하기

```python
age_median = diabetic["Age"].median()
```

* 나이를 오름차순으로 정렬했을 때, **가운데에 오는 값(중앙값)** 을 계산한다.
* 중앙값은 극단값(너무 어린 사람이나 너무 나이 많은 사람)의 영향을 덜 받기 때문에
  **“전형적인 나이 수준”을 나타내는 지표**로 많이 사용된다.

---

#### 6. 요약 결과 출력

```python
print("=== 당뇨 환자(Outcome=1) 연령 요약 ===")
print(f"최소 나이 : {age_min}")
print(f"최대 나이 : {age_max}")
print(f"평균 나이 : {age_mean:.2f}")
print(f"중앙값   : {age_median}")
```

* 첫 줄에서 제목을 출력:

  * `"=== 당뇨 환자(Outcome=1) 연령 요약 ==="`
* 그 다음 줄부터는 차례대로:

  * **최소 나이**: `age_min`
  * **최대 나이**: `age_max`
  * **평균 나이**: `age_mean` (소수 둘째 자리까지 `:.2f`로 출력)
  * **중앙값**: `age_median`
* 이 출력 결과를 통해,

  * 당뇨 환자 집단의 **연령 분포가 대략 어느 구간에 있는지**,
  * **평균과 중앙값이 얼마나 차이 나는지(왜곡 여부)** 등을 한눈에 파악할 수 있다.

---

### 한 줄 요약

> 이 코드는 **당뇨 환자만 따로 모은 뒤, 그들의 나이에 대해 최소값·최대값·평균·중앙값을 계산해서 “당뇨 환자 연령대가 어떤지” 요약해 주는 코드**이다.


### 미션 3. “BMI 높은 순 TOP 10 환자 뽑아주세요”

**업무 요청**

> “BMI가 높은 환자 10명을 ID, Age, BMI, Outcome만 보이도록 뽑아 주세요.”


In [ ]:
top10_bmi = df.sort_values(by="BMI", ascending = False)[["ID","Age","BMI","Outcome"]].head(10)
display(top10_bmi)

ID  Age   BMI  Outcome
474  TRAIN_474   22  57.3        0
430  TRAIN_430   26  55.0        1
579  TRAIN_579   28  52.9        1
625  TRAIN_625   23  52.3        0
359  TRAIN_359   36  50.0        1
598  TRAIN_598   31  49.7        1
331  TRAIN_331   26  49.6        1
51   TRAIN_051   37  48.8        1
463  TRAIN_463   32  48.3        1
296  TRAIN_296   43  47.9        1


---

### 코드 설명

```python
top10_bmi = df.sort_values(by="BMI", ascending=False)[["ID", "Age", "BMI", "Outcome"]].head(10)
from IPython.display import display
display(top10_bmi)
```

이 코드는 **BMI가 가장 높은 상위 10명의 정보를 뽑아서 보기 좋게 출력하는 코드**이다.

---

#### 1. BMI 기준으로 내림차순 정렬

```python
df.sort_values(by="BMI", ascending=False)
```

* `by="BMI"`
  → `BMI` 컬럼을 기준으로 정렬한다.
* `ascending=False`
  → **내림차순 정렬** (값이 큰 사람부터 위로 오도록 정렬)
* 결과:
  → BMI가 높은 사람 순서대로 정렬된 새로운 데이터프레임이 만들어진다.
  (원본 `df`는 그대로이고, 정렬된 결과가 반환된다)

---

#### 2. 필요한 컬럼만 선택

```python
[["ID", "Age", "BMI", "Outcome"]]
```

* 정렬된 데이터프레임에서

  * `ID`, `Age`, `BMI`, `Outcome` 4개 컬럼만 골라서 가져온다.
* 즉, BMI 상위 사람들에 대해:

  * **누구(ID)인지**
  * **나이(Age)는 몇 살인지**
  * **BMI 값은 얼마인지**
  * **당뇨 여부(Outcome)는 어떤지**
    핵심 정보만 추려 보여주기 위한 단계이다.

---

#### 3. 상위 10명만 추출

```python
.head(10)
```

* 정렬 + 컬럼 선택이 끝난 데이터프레임에서
  **위에서부터 10행만 가져온다.**
* `head(10)`은 “상위 10개 행만 보기”라는 뜻이다.
* 이렇게 해서 만들어진 결과를 `top10_bmi` 변수에 저장한다.

```python
top10_bmi = ...
```

* `top10_bmi`에는
  → **BMI가 가장 높은 상위 10명의 ID, 나이, BMI, Outcome 정보**가 담겨 있게 된다.

---

#### 4. display()로 보기 좋게 출력

```python
from IPython.display import display
display(top10_bmi)
```

* `display()` 함수는 코랩/주피터 노트북에서
  **판다스 데이터프레임을 표 형태로 예쁘게 출력**해준다.
* `print(top10_bmi)`보다

  * 컬럼 정렬, 줄번호, 테이블 형태가 잘 보이기 때문에
    **데이터 탐색 시 자주 사용하는 함수**이다.
* 이 코드로 BMI 상위 10명의 정보를
  눈으로 쉽게 확인할 수 있다.

---

### 한 줄 요약

> 이 코드는 **BMI가 높은 순으로 데이터를 정렬한 뒤, 그 중 상위 10명의 ID·나이·BMI·당뇨 여부만 골라서 표 형태로 출력**함으로써,
> “BMI가 특히 높은 사람들은 누구인지, 나이·Outcome과 어떤 관계가 있을지”를 탐색할 수 있게 해준다.


### 미션 4. “임신 횟수별 당뇨 발생 비율이 궁금합니다”

**업무 요청**

> “임신 횟수(`Pregnancies`)에 따라 당뇨병 발생률(Outcome=1)이 어떻게 달라지는지 보고 싶습니다.”


In [ ]:
p_o = df.groupby("Pregnancies")["Outcome"].mean()
print(p_o)

Pregnancies
0     0.333333
1     0.190476
2     0.200000
3     0.360656
4     0.338983
5     0.352941
6     0.325581
7     0.583333
8     0.533333
9     0.625000
10    0.380952
11    0.666667
12    0.428571
13    0.444444
14    1.000000
15    1.000000
Name: Outcome, dtype: float64



---

### 코드 설명

```python
pregnancy_outcome_rate = df.groupby("Pregnancies")["Outcome"].mean()
print(pregnancy_outcome_rate.sort_index())
```

이 코드는 **임신 횟수(Pregnancies)에 따라 당뇨(Outcome) 비율이 어떻게 달라지는지 확인하는 코드**이다.
(Outcome이 0/1 라벨이라고 가정하면, 평균값 = “당뇨일 확률/비율”로 볼 수 있음)

---

#### 1. 임신 횟수별로 그룹 나누기 + Outcome 평균 계산

```python
pregnancy_outcome_rate = df.groupby("Pregnancies")["Outcome"].mean()
```

* `df.groupby("Pregnancies")`

  * `Pregnancies` 값(예: 0회, 1회, 2회, …)이 같은 행끼리 묶어서 **그룹**을 만든다.
  * 즉, “임신 0번인 사람들 그룹”, “임신 1번인 사람들 그룹”, … 처럼 나뉜다고 보면 된다.
* `["Outcome"]`

  * 각 그룹에서 `Outcome` 컬럼만 선택한다.
* `.mean()`

  * 각 그룹에 대해 `Outcome` 값의 **평균**을 계산한다.

> 중요한 포인트
> Outcome이 0(비당뇨), 1(당뇨)인 이진값이라면,
> **`Outcome`의 평균 = 그 그룹에서 당뇨 환자 비율(= 당뇨일 확률)** 이 된다.

그래서 `pregnancy_outcome_rate`에는 예를 들면 이런 값이 들어간다:

* Pregnancies = 0 → Outcome 평균 = 0.10 (약 10%가 당뇨)
* Pregnancies = 1 → Outcome 평균 = 0.15 (약 15%가 당뇨)
* Pregnancies = 2 → Outcome 평균 = 0.22 …

---

#### 2. 임신 횟수 순서대로 정렬해서 출력

```python
print(pregnancy_outcome_rate.sort_index())
```

* `pregnancy_outcome_rate` 는 `Pregnancies` 값을 인덱스로 가지는 시리즈(Series)이다.
* `.sort_index()` 는

  * 인덱스(= Pregnancies 값)를 기준으로
  * **0, 1, 2, 3, … 순서대로 정렬**해준다.
* 그 결과를 `print()`로 출력하면:

  * 임신 0번일 때의 당뇨 비율
  * 임신 1번일 때의 당뇨 비율
  * 임신 2번일 때의 당뇨 비율
  * …
    을 **임신 횟수 증가 순서대로** 볼 수 있다.

---

### 한 줄 요약

> 이 코드는
> **임신 횟수(Pregnancies)별로 Outcome(당뇨 여부)의 평균을 구해서,
> ‘임신 경험이 많아질수록 당뇨 비율이 어떻게 변하는지’**
> 를 숫자로 확인할 수 있게 해주는 코드이다.


---
## 4. 마무리 정리

- CSV → DataFrame : `pd.read_csv`  
- 구조·타입 확인 : `head`, `shape`, `info`, `describe`  
- Pandas 기본 사용법 : DataFrame 생성, 인덱스/컬럼/행 선택, 조건 필터링, 정렬, 값 수정, 컬럼/행 추가·삭제  
- 결측치 확인 : `isnull().sum()`  
- 숨겨진 결측치(0) → `replace(0, np.nan)`  
- 결측치 처리 : `dropna`, `fillna`(평균/중앙값/최빈값 등)  
- Seaborn으로 샘플 데이터(`titanic`)를 불러와 구조·결측치를 확인하고, 간단한 그래프를 그릴 수 있다.